dali/pipeline/operators/expressions/arithmetic.h

```diff
@@ -128,50 +132,54 @@ struct ExprImplTask {
 };
 
 template <typename Backend>
-inline void CreateExecutionOrder(std::vector<ExprImplTask> &order, const ExprNode &expr,
-                                 ExprImplCache &cache) {
+inline void CreateExecutionTasks(std::vector<ExprImplTask> &order, const ExprNode &expr,
+                                 ExprImplCache &cache, cudaStream_t stream) {
   if (expr.GetNodeType() != NodeType::Function) {
     return;
   }
-  auto &func = dynamic_cast<const ExprFunc&>(expr);
+  auto &func = dynamic_cast<const ExprFunc &>(expr);
   for (int i = 0; i < expr.GetSubexpressionCount(); i++) {
-    CreateExecutionOrder<Backend>(order, func[i], cache);
+    CreateExecutionTasks<Backend>(order, func[i], cache, stream);
   }
-  order.push_back({cache.GetExprImpl<Backend>(func), {&func}});
+  order.push_back({cache.GetExprImpl<Backend>(func), {stream, &func}});
 }
 
 template <typename Backend>
-inline std::vector<ExprImplTask> CreateExecutionOrder(const ExprNode &expr,
-                                                            ExprImplCache &cache) {
+inline std::vector<ExprImplTask> CreateExecutionTasks(const ExprNode &expr, ExprImplCache &cache,
+                                                      cudaStream_t stream) {
   std::vector<ExprImplTask> result;
-  CreateExecutionOrder<Backend>(result, expr, cache);
+  CreateExecutionTasks<Backend>(result, expr, cache, stream);
   return result;
 }
 
-inline kernels::TensorListShape<> ShapePromotion(std::string op,
-                                                 const kernels::TensorListShape<> &left,
-                                                 const kernels::TensorListShape<> &right) {
-  bool is_left_scalar = IsScalarLike(left);
-  bool is_right_scalar = IsScalarLike(right);
+inline kernels::TensorListShape<> ShapePromotion(
+    std::string op, const std::vector<kernels::TensorListShape<>> &shapes) {
+  assert(shapes.size() == 1 || shapes.size() == 2 && "Not implemented for other arity");
+  if (shapes.size() == 1) {
+    return shapes[0];
+  }
+  bool is_left_scalar = IsScalarLike(shapes[0]);
+  bool is_right_scalar = IsScalarLike(shapes[1]);
   if (is_left_scalar && is_right_scalar) {
     return kernels::TensorListShape<>{{1}};
   }
   if (is_left_scalar) {
-    return right;
+    return shapes[1];
   }
   if (is_right_scalar) {
-    return left;
+    return shapes[0];
   }
   using std::to_string;
-  DALI_ENFORCE(left == right, "Input shapes of element-wise arithemtic operator \"" + op +
-                                  "\" do not match. Expected equal shapes, got: " + op + "(" +
-                                  to_string(left) + ", " + to_string(right) + ").");
-  return left;
+  DALI_ENFORCE(shapes[0] == shapes[1], "Input shapes of elemenetwise arithemtic operator \"" + op +
```

Comment:
Nitpick: the `make_string` utility is already merged - you'd avoid all those `+` and `to_string`.

Reply:
Done